bin/morse

```diff
@@ -63,76 +28,148 @@ my %translations = reverse (
-        while (/,\n/ && defined ($line = <FILE>)) {
-            $line =~ s/^\s+//;          # Leading space.
-            $line =~ s/dit ?/./g;       # Long to short.
-            $line =~ s/daw ?/-/g;       # Long to short.
-            s/,\n/ $line/;              # Combine lines.
+my %opt;
+getopts('ds', \%opt) or die("usage: $0 [-ds] [string ...]\n");
+if (@ARGV) {
+    if ($opt{'d'}) {
+        $opt{'s'} ? decode_dot_arg() : decode_word_arg();
+    } else {
+        $opt{'s'} ? encode_dot_arg() : encode_word_arg();
+    }
+} else {
+    if ($opt{'d'}) {
+        $opt{'s'} ? decode_dot() : decode_word();
+    } else {
+        $opt{'s'} ? encode_dot() : encode_word();
+    }
+}
+
+sub mors2txt {
+    my $line = shift;
+
+    foreach my $m (split /\s+/, $line) {
+        if (!exists($translations{$m})) {
+            die "$m: unknown token";
         }
-        my $comma = s/,$//;             # Trailing comma.
-        s/([-.]+) ?/exists $translations{$1} ?
-                           $translations{$1} : die "$1: unknown token"/ge;
-        print;
-        last if $comma;
+        print $translations{$m};
     }
-    close FILE or die "$file: $!\n";
+    print "\n" if (tell(*STDOUT) > 0);
 }
 
-exit $warnings;
+sub txt2mors {
+    my $line = shift;
+
+    foreach my $c (split //, $line) {
+        if ($c eq "\n" || $c eq ' ') {
+            print "\n";
+        } elsif (exists $chartab{$c}) {
+            print ($chartab{$c}, "\n");
+        }
+    }
+}
+
+sub txt2word {
+    my $line = shift;
+
+    foreach my $c (split //, $line) {
+        if ($c eq "\n") {
+            print "\n";
+        } elsif (exists $chartab{$c}) {
+            my $w = $chartab{$c};
+            $w =~ s/\./dit /g;
+            $w =~ s/\-/daw /g;
+            print ($w, "\n");
+        }
+    }
+}
+
+sub word2txt {
+    my $line = shift;
+
+    foreach my $part (split /[,\n]/, $line) {
+        my $m = '';
+        foreach my $tok (split /\s+/, $part) {
+            next if (length($tok) == 0);
+            if (!exists($wordtab{$tok})) {
+                die "$tok: unknown token";
+            }
+            $m .= $wordtab{$tok};
+        }
+        next if (length($m) == 0);
+        if (!exists($translations{$m})) {
+            die "$m: unknown token";
+        }
+        print $translations{$m};
+    }
+    print "\n" if (tell(*STDOUT) > 0);
+}
+
+sub decode_dot_arg  { mors2txt(join "\n", @ARGV) }
+sub decode_dot      { mors2txt(join '', readline) }
+sub decode_word_arg { word2txt(join ' ', @ARGV) }
+sub decode_word     { word2txt(join "\n", readline) }
+sub encode_dot_arg  { txt2mors(join "\n", @ARGV) }
+sub encode_dot      { txt2mors(join '', readline) }
+sub encode_word_arg { txt2word(join "\n", @ARGV) }
+sub encode_word     { txt2word(join '', readline) }
 
 __END__
 
 =pod
 
 =head1 NAME
 
-morse - read morse and translate it to text
+morse -  translate text to morse code
 
 =head1 SYNOPSIS
 
-morse [files ...]
+morse [-ds] [string ...]
```

Comment:
What about the case of reading from files, as in the original? The bad version doesn't do files, but this is a big change from the PPT version.

Reply:
Possibly we can support both by adding a "files" flag (-f ?), where ARGV is treated as a file list instead of a string list. Then the original PPT behaviour is not lost.

Reply:
that seems fine